parsec/backend/client_context.py

```diff
@@ -27,13 +27,21 @@
 
 
 class BaseClientContext:
-    __slots__ = ("conn_id", "api_version")
+    __slots__ = ("conn_id", "api_version", "_cancel_scope")
     TYPE: ClientType
     logger: BoundLogger
 
     def __init__(self, api_version: ApiVersion):
         self.api_version = api_version
         self.conn_id = uuid4().hex
+        self._cancel_scope: Optional[trio.CancelScope] = None
+
+    def set_scope(self, client_scope: trio.CancelScope) -> None:
+        self._cancel_scope = client_scope
+
+    def exit_cancel_scope(self) -> None:
```

Comment:
Good idea have a method that does "close the client's cancel scope if any"

But I think we should not have an assert here but just do nothing if there is no cancel scope. I theory this assert is a sanity check, but I suspect the code is complex enough that we will use this method in a callback that is also used in `rpc.py` (where there is no cancel scope) and create unexpected errors.

So the idea would be to rename this method something like `close_connection_asap`, which means we will try to close the connection if possible, and otherwise let it finish.
This is good enough in practice given the connections in rpc.py don't have a cancel scope because they are short lived (the execute a single command, unlike the websocket connections that stay open and execute multiple commands)